File: examples/ruby/product/lib/product_logic.rb
# frozen_string_literal: true

require 'time'
require_relative 'command_validation_error'

# Pure business logic for product aggregate.
# No gRPC dependencies - can be tested in isolation.
module ProductLogic
  ProductState = Struct.new(:sku, :name, :description, :price_cents, :status, keyword_init: true) do
    def initialize(sku: '', name: '', description: '', price_cents: 0, status: :uninitialized)
      super
    end

    def exists?
      status != :uninitialized
    end

    def active?
      status == :active
    end
  end

  class << self
    def rebuild_state(event_book)
      state = ProductState.new

      return state if event_book.nil? || event_book.pages.empty?

      # Start from snapshot if present
      if event_book.snapshot&.state
        snap = event_book.snapshot.state
        if snap.type_url.end_with?('ProductState')
          snap_state = Examples::ProductState.decode(snap.value)
          state = ProductState.new(
            sku: snap_state.sku,
            name: snap_state.name,
            description: snap_state.description,
            price_cents: snap_state.price_cents,
            status: snap_state.status.to_sym
          )
        end
      end

      # Apply events
      event_book.pages.each do |page|
        next unless page.event

        state = apply_event(state, page.event)
      end

      state
    end

    def handle_create_product(state, sku, name, description, price_cents)
      raise CommandValidationError.failed_precondition('Product already exists') if state.exists?
      raise CommandValidationError.invalid_argument('SKU is required') if sku.nil? || sku.empty?
      raise CommandValidationError.invalid_argument('Name is required') if name.nil? || name.empty?
      raise CommandValidationError.invalid_argument('Price must be positive') if price_cents <= 0

      Examples::ProductCreated.new(
        sku: sku,
        name: name,
        description: description || '',
        price_cents: price_cents,
        created_at: now_timestamp
      )
    end

    def handle_update_product(state, name, description)
      raise CommandValidationError.failed_precondition('Product does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Product is discontinued') unless state.active?

      Examples::ProductUpdated.new(
        name: name || state.name,
        description: description || state.description
      )
    end

    def handle_set_price(state, price_cents)
      raise CommandValidationError.failed_precondition('Product does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Product is discontinued') unless state.active?
      raise CommandValidationError.invalid_argument('Price must be positive') if price_cents <= 0

      Examples::PriceSet.new(
        old_price_cents: state.price_cents,
        new_price_cents: price_cents
      )
    end

    def handle_discontinue(state, reason)
      raise CommandValidationError.failed_precondition('Product does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition('Product already discontinued') unless state.active?

      Examples::ProductDiscontinued.new(
        reason: reason || '',
        discontinued_at: now_timestamp
      )
    end

    private

    def apply_event(state, event_any)
      type_url = event_any.type_url

      if type_url.end_with?('ProductCreated')
        event = Examples::ProductCreated.decode(event_any.value)
        ProductState.new(
          sku: event.sku,
          name: event.name,
          description: event.description,
          price_cents: event.price_cents,
          status: :active
        )
      elsif type_url.end_with?('ProductUpdated')
        event = Examples::ProductUpdated.decode(event_any.value)
        ProductState.new(
          sku: state.sku,
          name: event.name,
          description: event.description,
          price_cents: state.price_cents,
          status: state.status
        )
      elsif type_url.end_with?('PriceSet')
        event = Examples::PriceSet.decode(event_any.value)
        ProductState.new(
          sku: state.sku,
          name: state.name,
          description: state.description,
          price_cents: event.new_price_cents,
          status: state.status
        )
      elsif type_url.end_with?('ProductDiscontinued')
        ProductState.new(
          sku: state.sku,
          name: state.name,
          description: state.description,
          price_cents: state.price_cents,
          status: :discontinued
        )
      else
        state
      end
    end

    def now_timestamp
      Google::Protobuf::Timestamp.new(seconds: Time.now.to_i)
    end
  end
end

File: examples/cpp/product/src/product_logic.cc
#include "product_logic.hpp"
#include "validation_error.hpp"
#include <chrono>

namespace product {

using namespace angzarr;

ProductState ProductLogic::rebuild_state(const angzarr::EventBook* event_book) {
    ProductState state;
    if (!event_book || event_book->pages().empty()) return state;

    for (const auto& page : event_book->pages()) {
        if (page.has_event()) {
            state = apply_event(std::move(state), page.event());
        }
    }
    return state;
}

examples::ProductCreated ProductLogic::handle_create_product(
    const ProductState& state, const std::string& sku, const std::string& name,
    const std::string& description, int32_t price_cents) {
    if (state.exists()) throw ValidationError::failed_precondition("Product already exists");
    if (sku.empty()) throw ValidationError::invalid_argument("SKU is required");
    if (name.empty()) throw ValidationError::invalid_argument("Name is required");
    if (price_cents <= 0) throw ValidationError::invalid_argument("Price must be positive");

    examples::ProductCreated event;
    event.set_sku(sku);
    event.set_name(name);
    event.set_description(description);
    event.set_price_cents(price_cents);
    event.mutable_created_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::ProductUpdated ProductLogic::handle_update_product(
    const ProductState& state, const std::string& name, const std::string& description) {
    if (!state.exists()) throw ValidationError::failed_precondition("Product does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Product is discontinued");

    examples::ProductUpdated event;
    event.set_name(name.empty() ? state.name : name);
    event.set_description(description.empty() ? state.description : description);
    return event;
}

examples::PriceSet ProductLogic::handle_set_price(const ProductState& state, int32_t price_cents) {
    if (!state.exists()) throw ValidationError::failed_precondition("Product does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Product is discontinued");
    if (price_cents <= 0) throw ValidationError::invalid_argument("Price must be positive");

    examples::PriceSet event;
    event.set_old_price_cents(state.price_cents);
    event.set_new_price_cents(price_cents);
    return event;
}

examples::ProductDiscontinued ProductLogic::handle_discontinue(
    const ProductState& state, const std::string& reason) {
    if (!state.exists()) throw ValidationError::failed_precondition("Product does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Product already discontinued");

    examples::ProductDiscontinued event;
    event.set_reason(reason);
    event.mutable_discontinued_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

ProductState ProductLogic::apply_event(ProductState state, const google::protobuf::Any& event) {
    const auto& type_url = event.type_url();

    if (type_url.find("ProductCreated") != std::string::npos) {
        examples::ProductCreated e;
        event.UnpackTo(&e);
        state.sku = e.sku();
        state.name = e.name();
        state.description = e.description();
        state.price_cents = e.price_cents();
        state.status = ProductStatus::Active;
    } else if (type_url.find("ProductUpdated") != std::string::npos) {
        examples::ProductUpdated e;
        event.UnpackTo(&e);
        state.name = e.name();
        state.description = e.description();
    } else if (type_url.find("PriceSet") != std::string::npos) {
        examples::PriceSet e;
        event.UnpackTo(&e);
        state.price_cents = e.new_price_cents();
    } else if (type_url.find("ProductDiscontinued") != std::string::npos) {
        state.status = ProductStatus::Discontinued;
    }

    return state;
}

}  // namespace product
